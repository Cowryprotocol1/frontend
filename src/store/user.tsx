import React, { createContext, useContext, useEffect,useState, useCallback } from "react";

import { User } from "@/constant/dummydata";


const url = `https://cowryprotocol.io`;
const stellar_url = ` https://horizon-testnet.stellar.org`



export type UserProp = {
  id: number;
  name: string;
  image: string;
  wallet_address: string;
  role: "user" | "ifp";
  joined: Date;
};

export type UserContextProps = {
  userData: UserProp | null;
  setUserData: React.Dispatch<React.SetStateAction<UserProp>>;
  walletAddress: "";
  setWalletAddress: React.Dispatch<React.SetStateAction<any>>;
  walletVendor: "";
  setWalletVendor: React.Dispatch<React.SetStateAction<any>>;
  IFPData: any; 
  setIFPData:React.Dispatch<React.SetStateAction<any>>;
  logout: boolean;
  setLogout:  React.Dispatch<React.SetStateAction<boolean>>;
  toggleLogoutMode: any;
  getTransactions: any;
  getBalance: any;
  getDepositIntent: any;
  getAccount: any;
  getWithdrawalIntent: any;
  onboardIFP:any;
  postPaymentConfirmation: any;
  role: string;
  getTransactionStatus: any;
  signXDR: any;
  

  getDepositIntentIFP: any;
  transactions: any; 
  setTransactions: React.Dispatch<React.SetStateAction<any>>;
  setBalances:React.Dispatch<React.SetStateAction<any>>;
  balances: any;
  depositOpen:any;
  setDepositOpen:React.Dispatch<React.SetStateAction<boolean>>;
  withdrawOpen:any;
  setWithdrawOpen: React.Dispatch<React.SetStateAction<boolean>>;
  conversionOpen:any;
  setConversionOpen:React.Dispatch<React.SetStateAction<boolean>>;
  setRole:React.Dispatch<React.SetStateAction<string>>;

};

const UserContext = createContext<UserContextProps>({
  userData: User[0] || null,
  setUserData: () => null,

  logout: null ,
  setLogout: ()=> null,
  toggleLogoutMode: ()=>null,
  getTransactions: ()=>null,
  getBalance: ()=>null,
  getWithdrawalIntent:()=>null,
  onboardIFP:()=>null,
  role: "",
  getDepositIntentIFP:()=>null,
  getTransactionStatus:()=>null,
  signXDR:()=> null,

  walletAddress: "",
  setWalletAddress: () => null,
  walletVendor: "",
  setWalletVendor: () => null,

  transactions: null,
  setTransactions: ()=> null,
  balances: null,
  setBalances: ()=> null,
  getDepositIntent: ()=> null,
  depositOpen: null,
  setDepositOpen:()=>null,
  withdrawOpen: null,
  setWithdrawOpen: ()=>null,
  postPaymentConfirmation:()=>null,
  conversionOpen:null,
  setConversionOpen: ()=>null,
  IFPData: null, 
  setRole:()=>null,
  setIFPData:()=>null,
  getAccount:()=>null,
});



const initialRoleStatus = () => {
  if (typeof window === "undefined") return;
  const role = localStorage.getItem("userType") 
  return role

}


export const UserProvider = ({ children }: { children: React.ReactNode }) => {
  // TODO: data is expected to obtained from api endpoint
  const [userData, setUserData] = useState(null);
  const [logout, setLogout] = useState(false)
  const [role, setRole] = useState(initialRoleStatus)
  const [walletAddress, setWalletAddress] = useState(null);
  const [walletVendor, setWalletVendor] = useState(null);
  const [transactions, setTransactions] = useState([]);
  const [balances, setBalances] = useState([]);
  const [depositOpen, setDepositOpen] = useState(false);
  const [withdrawOpen, setWithdrawOpen] = useState(false);
  const [conversionOpen, setConversionOpen] = useState(false)
  const [IFPData, setIFPData] = useState(null)

  const toggleLogoutMode = () => {
    if (logout !== null) {
      setLogout(true);
      localStorage.removeItem("logout")
      localStorage.removeItem("userType")
      localStorage.removeItem("walletAddress")
      window.location.href = "/";
    } 
    return logout;
  
  };

  const getDepositIntent = async (data: any)=> {
    let rData = {
      amount: parseFloat(data.amount),
      blockchainAddress: data.address,
      bankType: data.bank,
      narration: data.description
    }
    // console.log(JSON.stringify(rData))
    try {
      const response = await fetch(`${url}/deposit`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(rData),
      })
      let res = response.json()
      return res
    } catch (error) {
      throw error;
    }
  };
  const postPaymentConfirmation = async (data: any)=> {
    // console.log(JSON.stringify(data))
    try {
      const response = await fetch(`${url}/deposit`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(data),
      })
      let res = response.json()
      return res
    } catch (error) {
      throw error;
    }
  };
  
  const onboardIFP = async (data: any)=> {
    let wData = {
      blockchainAddress: walletAddress,
      bankName: data.bank_name,
      bankAccount: data.account_number,
      email: data.email,
      phoneNumber: data.phone,
    }
    // console.log(wData)
    try {
      const response = await fetch(`${url}/onboard`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(wData),
      })
      let res = response.json()
      return res
    } catch (error) {
      throw error;
    }
  };

  const getDepositIntentIFP = async (txn_id: string)=> {
    let wData = {
      merchant_pubKey: walletAddress,
      transaction_Id: txn_id,
      merchant_id: IFPData?.account_id,

    }
    // console.log(wData)
    try {
      const response = await fetch(`${url}/merchants`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(wData),
      })
      let res = response.json()
      return res
    } catch (error) {
      throw error;
    }
  };


  const getWithdrawalIntent = async (data: any)=> {
    let wData = {
      amount: parseFloat(data.amount),
      blockchain_address: data.address,
      bank_name: data.bank,
      account_number: data.account_number,
      name_on_acct: data.account_name,
      phone_number: data.phone,
      transaction_narration: data.description
    }
    // console.log(wData)
    try {
      const response = await fetch(`${url}/withdrawal`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(wData),
      })
      let res = response.json()
      return res
    } catch (error) {
      throw error;
    }
  };
  
  const getTransactions = async (address: string, role: string)=> {
    try {
      const response = await fetch(`${url}/merchants?public_key=${address}&query_type=${role}`)
      let res = response.json()
      return res
    } catch (error) {
      throw error;
    }
  };
  const getBalance = async (address: string)=> {
    try {
      const response = await fetch(`${stellar_url}/accounts/${address}`)
      let res = response.json()
      return res
    } catch (error) {
      throw error;
    }
  };

  const getAccount = async (address: string)=> {
    try {
      const response = await fetch(`${url}/accounts?account_id=${address}`)
      let res = response.json()
      return res
      
    } catch (error) {
      throw error;
    }
  };


  

  const signXDR = async (signed_xdr: string) => {
    // console.log(transactionId)
    let xdrData = {
      signed_xdr: signed_xdr,
     
    }
    // console.log(transactionData)
    try {
      const response = await fetch(`${url}/submit_xdr`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(xdrData),
      })
      let res = response.json()
      console.log(res)
      return res

    } catch (error) {
      console.log(error)
      throw error;
    }
  };
 
  
const getTransactionStatus = async (transactionId: string) => {
    // console.log(transactionId)
    let transactionData = {
      transactionId: transactionId,
     
    }
    // console.log(transactionData)
    try {
      const response = await fetch(`${url}/transactionStatus`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(transactionData),
      })
      let res = response.json()
      // console.log(res)
      return res

    } catch (error) {
      console.log(error)
      throw error;
    }
  };
  
  return (
    <UserContext.Provider value={{
      userData, 
      setUserData, 

      walletAddress, 
      setWalletAddress, 
      walletVendor, 
      setWalletVendor, 
      logout,
      setLogout,
      toggleLogoutMode,
      getTransactions,
      getBalance,
      onboardIFP,
      signXDR,
      role,
      setRole,
      transactions,
      setTransactions,
      balances,
      setBalances,
      getDepositIntent,
      getWithdrawalIntent,
      depositOpen,
      setDepositOpen,
      withdrawOpen,
      setWithdrawOpen,
      postPaymentConfirmation,
      conversionOpen,
      setConversionOpen,
      IFPData,
      setIFPData,
      getAccount,
      getDepositIntentIFP,
      getTransactionStatus,
    }}>
      {children}
    </UserContext.Provider>
  );
};

export const useUser = () => useContext(UserContext);
