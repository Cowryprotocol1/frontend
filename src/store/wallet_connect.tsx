
import { StellarWalletsKit, WalletNetwork, WalletType } from 'stellar-wallets-kit';

 export const kit = new StellarWalletsKit({
    network: WalletNetwork.TESTNET,
    selectedWallet: WalletType.FREIGHTER
  });

