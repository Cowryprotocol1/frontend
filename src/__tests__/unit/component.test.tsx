import { render, screen, fireEvent } from "@testing-library/react";
// import {build, fake } from '@jackfranklin/test-data-bot';
import 'intersection-observer'; 
import { UserProvider } from "@/store/user";
import WithdrawModal from "@/components/balance/withdrawModal";
import DepositModal from "@/components/balance/depositModal";
import CountdownTimer from "@/components/balance/timer";

import Donut from "@/components/donut";

import Avatar from "@/components/icons/avatar";
const defaultWDProps = {
    timer: 30,
    isOpen: true,
    NGN: [{balance:"3000"}],
    setModalOpen: jest.fn(),
  };
const defaultCountDownProps = {
    timer: 30,
    setIsExpired: jest.fn()
}
const defaultAvatarProps = {
  src: "https://picsum.photos/200",
  alt: "alt_image",
  width: 70,
  height: 70,
  className: "",
  editBg: "https://picsum.photos/200",
  onClick: jest.fn()
}

// Avatar Component Test
describe('Avatar', () => {
    it('❌ renders Component correctly', () => { 
        const { asFragment } = render(<Avatar {...defaultAvatarProps}/>);
        expect(asFragment()).toMatchSnapshot();
    });
});

// Donut Component Test
describe('Donut', () => {
    it('❌ renders Component correctly', () => { 
        const { asFragment } = render(
            <UserProvider>
        <Donut />
    </UserProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});


// CountDown Component Test
describe('CountdownTimer', () => {
    it('❌ renders Component correctly', () => { 
        const { asFragment } = render(<CountdownTimer {...defaultCountDownProps} />);
        expect(asFragment()).toMatchSnapshot();
    });
});


// Deposit Component Test
describe("DepositModal ", () => {

  const depositComponent = ()=> render(
    <UserProvider>
        <DepositModal {...defaultWDProps} />
    </UserProvider>
  );

  it("❌ renders Component correctly", () => {
    const { asFragment } = depositComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  it(" ❌ count number of input in the form'", async() => {
    const {getAllByTestId} = depositComponent();
    const formInputs = getAllByTestId("form");
    expect(await formInputs).toHaveLength(4);
  });

  it(" ❌ Submit Form Data'", async() => {
    const {getByTestId, getByText, getAllByTestId, getByPlaceholderText} = depositComponent();

    const amount: any = getByPlaceholderText("Deposit Amount");
    fireEvent.change(amount, { target: { value: "1000" } });
    expect(amount.value).toBe("1000");

    const bank: any = getByPlaceholderText("Deposit Bank Name");
    fireEvent.change(bank, { target: { value: "GTB" } });
    expect(bank.value).toBe("GTB");

    const desc: any = getByPlaceholderText("Wallet Address");
    fireEvent.change(desc, { target: { value: "GRXISJSJ2DJ4H2N44NK2N2" } });
    expect(desc.value).toBe("GRXISJSJ2DJ4H2N44NK2N2");

    const accNum: any = getByPlaceholderText("Transaction Description");
    fireEvent.change(accNum, { target: { value: "TEST DESCRIPTION" } });
    expect(accNum.value).toBe("TEST DESCRIPTION");

    const button = getByTestId('submit_btn')
    fireEvent.click(button);
    expect(await getByText('Confirm Deposit')).toBeInTheDocument();
    // check for form two if it appears
    const formInputs = getAllByTestId("form_two");
    expect(await formInputs).toHaveLength(6);
   });

  
});

// Withdrwal Component Test

describe("Withdrawal Component", () => {
    // defining the component
    const testComponent = ()=> render(
        <UserProvider>
            <WithdrawModal {...defaultWDProps} />
        </UserProvider>
      );
  
    it(" ❌ renders Withdrawal Component correctly'", () => {
     const {asFragment} = testComponent();
     expect(asFragment()).toMatchSnapshot();
    });
  
    it(" ❌ count number of input in the form'", async () => {
      let component = testComponent();
      const formInputs = component.getAllByTestId('form')
      expect(await formInputs).toHaveLength(6);
     });

     it(" ❌ Submit Form Data'", async() => {
      const {getByTestId, getByText, getAllByTestId, getByPlaceholderText} = testComponent();
        // fill the form
        const amount:any = getByPlaceholderText('Withdraw Amount')
        fireEvent.change(amount,{target: {value: '1000'}})
        expect(amount.value).toBe('1000');

        const bank:any = getByPlaceholderText('Bank Name')
        fireEvent.change(bank,{target: {value: 'GTB'}})
        expect(bank.value).toBe('GTB');

        const desc:any = getByPlaceholderText('Transaction Description')
        fireEvent.change(desc,{target: {value: 'Test Desc'}})
        expect(desc.value).toBe('Test Desc');

        const accNum:any = getByPlaceholderText('Account Number')
        fireEvent.change(accNum,{target: {value: '0987654321'}})
        expect(accNum.value).toBe('0987654321');

        const accName:any = getByPlaceholderText('Account Name')
        fireEvent.change(accName,{target: {value: 'Daniel Olagunju'}})
        expect(accName.value).toBe('Daniel Olagunju');

        const phone:any = getByPlaceholderText('Phone Number');
        fireEvent.change(phone,{target: {value: '08012345678'}});
        expect(phone.value).toBe('08012345678');

        const button = getByTestId('submit_btn')
        fireEvent.click(button);
        expect(await getByText('Confirm Withdrawal')).toBeInTheDocument();
      // check for form two if it appears
        const formInputs = getAllByTestId("form_two");
        expect(await formInputs).toHaveLength(6);
     });
  });
