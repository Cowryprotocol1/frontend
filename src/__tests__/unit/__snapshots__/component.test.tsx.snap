// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar ❌ renders Component correctly 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col justify-center items-center"
  >
    <img
      alt="alt_image"
      class=""
      height="70"
      src="https://picsum.photos/200"
      width="70"
    />
    <span
      class="https://picsum.photos/200 -mt-8 ml-4 p-2 rounded-full"
    >
      <svg
        class="text-white_day"
        fill="currentColor"
        height="15"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512 144v288c0 26.5-21.5 48-48 48H48c-26.5 0-48-21.5-48-48V144c0-26.5 21.5-48 48-48h88l12.3-32.9c7-18.7 24.9-31.1 44.9-31.1h125.5c20 0 37.9 12.4 44.9 31.1L376 96h88c26.5 0 48 21.5 48 48zM376 288c0-66.2-53.8-120-120-120s-120 53.8-120 120 53.8 120 120 120 120-53.8 120-120zm-32 0c0 48.5-39.5 88-88 88s-88-39.5-88-88 39.5-88 88-88 88 39.5 88 88z"
        />
      </svg>
    </span>
  </div>
</DocumentFragment>
`;

exports[`CountdownTimer ❌ renders Component correctly 1`] = `
<DocumentFragment>
  <div
    class="bg-[#EBA352] h-[50px] w-[50px] rounded-full flex flex-col justify-center items-center"
  >
    <h2
      class="text-xs font-thin text-white"
    >
      30:00
    </h2>
  </div>
</DocumentFragment>
`;

exports[`DepositModal  ❌ renders Component correctly 1`] = `
<DocumentFragment>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</DocumentFragment>
`;

exports[`Donut ❌ renders Component correctly 1`] = `
<DocumentFragment>
  <div
    class="px-8 flex flex-row items-center justify-center"
  >
    <canvas
      height="230"
      role="img"
      width="230"
    />
  </div>
</DocumentFragment>
`;

exports[`Withdrawal Component  ❌ renders Withdrawal Component correctly' 1`] = `
<DocumentFragment>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</DocumentFragment>
`;
